File: esp/lib/uart/espController.cpp
#include "EspController.h"

#include "Protocol.h"
#include "Commands.h"



EspController::EspController(
    UartHandler &uart,
    TextStorage &storage
)
:
_uart(uart),
_storage(storage)
{

}

void EspController::begin() {
    _storage.begin();

}

void EspController::loop()
{

    if(!_uart.available()) {
        return;
    }

    String message =_uart.readLine();
    String command;
    String data;
    if(Protocol::decode(message,command,data)) {
        handleMessage(command, data);
    }
}

void EspController::requestText(){
    send(Commands::GET, Commands::TEXT);
}

void EspController::handleMessage(String command, String data) {
    /*
       Raspberry Pi Antwort:
       TEXT|Mein langer Text
    */
    if(command == Commands::TEXT) {
        if(_storage.save(data)) {
            send(Commands::SET,"TEXT_OK");
        }
        else {
            send(Commands::SET, "TEXT_ERROR");
        }
    }
    /* SET Anfrage */
    else if(command == Commands::SET) {
        send(Commands::SET, "OK");
    }
}

void EspController::send(String command, String data)
{
    String message = Protocol::encode(command, data);
    _uart.send(message);
}

String EspController::getText() {
    return _storage.load();
}

bool EspController::hasText() {
    return _storage.exists();
}

File: rpi/tests/uart/esp_controller.py
"""
esp_controller.py

Raspberry Pi Controller

Beantwortet Anfragen
vom ESP8266.
"""


from uart import UART
from uart_switch import UARTSwitch
from protocol import Protocol
from commands import Commands



class ESPController:


    def __init__(self, text_provider, config_provider=None, port="/dev/serial0", baudrate=115200, switch_gpio=23):

        self.text_provider = text_provider
        self.config_provider = config_provider

        self.switch = UARTSwitch(
            gpio=switch_gpio
        )

        self.uart = UART(
            port=port,
            baudrate=baudrate
        )


    def loop(self):
        self.switch.enable()
        try:
            if self.uart.available():
                message = self.uart.readline()
                command, data = Protocol.decode(
                    message
                )
                self.handle(
                    command,
                    data
                )

        finally:
            self.switch.disable()



    def handle(self, command, data):
        # ESP fordert Daten an
        if command == Commands.GET:
            # Text senden
            if data == Commands.TEXT:
                text = self.text_provider()
                self.send(Commands.TEXT,text)

            # Konfiguration senden
            elif data == Commands.CONFIG:
                if self.config_provider:
                    config = self.config_provider()
                    self.send(
                        Commands.CONFIG,
                        config
                    )


        # ESP sendet Daten
        elif command == Commands.SET:
            self.handle_set(
                data
            )



    def handle_set(self, data):
        """
        Verarbeitung von
        SET Nachrichten

        Beispiel:
        SET|CONFIG
        """

        self.send(
            Commands.OK,
            data
        )



    def send(self, command, data=""):
        message = Protocol.encode(
            command,
            data
        )
        self.switch.enable()
        try:
            self.uart.write(
                message
            )
        finally:
            self.switch.disable()



    def close(self):
        self.uart.close()
        self.switch.close()